File: Src/FUSE/Linux/x86-64/I-O_Engine/check.c
/*
 * check.c — cafs_check(). Verifies the Superblock's BLAKE3-128
 * checksum. On mismatch, tries Superblock Backup (read at mount() via
 * LBA0's fallback pointer); if the backup checks out, copies it over
 * the primary Superblock in place, fsyncs, and logs the repair to
 * stderr — "simple and foreseeable," per the locked V1 scope, so this
 * fixes rather than just reports. All other anchor structures are out
 * of scope for V1 ("ignore all anchor block errors except
 * Superblock") — read best-effort by mount(), not verified here.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 * Copyright (c) 2026, Imran Bin Gifary (System Delta)
 */
#include <stdio.h>
#include <string.h>
#include "cafs_io_internal.h"
#include "syscalls.h"

cafs_status_t cafs_check(cafs_ctx_t *ctx, cafs_check_result_t *out_result)
{
    if (!ctx || !out_result) return CAFS_ERR_INVALID_ARG;
    if (!ctx->anchor_loaded) return CAFS_ERR_NOT_MOUNTED;

    memset(out_result, 0, sizeof(*out_result));
    out_result->pointer_checksum_algo_used = ctx->lba0_pointer_algo;

    pthread_mutex_lock(&ctx->lock);

    int primary_ok = cafs_verify_blake3_region(ctx->superblock, SB_HASHED_LEN, SB_OFF_CHECKSUM);

    if (primary_ok) {
        out_result->superblock_ok = 1;
    } else if (ctx->superblock_backup_valid &&
               cafs_verify_blake3_region(ctx->superblock_backup, SB_HASHED_LEN, SB_OFF_CHECKSUM)) {
        fprintf(stderr,
                "cafs: primary Superblock checksum mismatch at offset %llu; "
                "repairing from Superblock Backup at offset %llu\n",
                (unsigned long long)ctx->lba0.real_lba,
                (unsigned long long)ctx->lba0.fallback_real_lba);

        memcpy(ctx->superblock, ctx->superblock_backup, sizeof(ctx->superblock));

        if (ctx->want_write) {
            cafs_status_t wst = cafs_write_raw(ctx, ctx->lba0.real_lba,
                                                CAFS_REAL_BLOCK_4K, ctx->superblock);
            if (wst == CAFS_OK)
                cafs_sys_fdatasync(ctx->fd);
            else
                fprintf(stderr, "cafs: repair copy succeeded in memory but the write-back "
                                 "to the primary Superblock failed (mount is still usable "
                                 "this session; the repair did not persist)\n");
        } else {
            fprintf(stderr, "cafs: mount is read-only; repair applied in memory only, "
                             "not written back\n");
        }

        out_result->superblock_ok = 1;
        out_result->superblock_repaired = 1;
    } else {
        fprintf(stderr, "cafs: primary Superblock checksum mismatch and no usable backup; "
                         "reporting failure, not crashing\n");
        out_result->superblock_ok = 0;
    }

    out_result->overall_ok = out_result->superblock_ok;

    pthread_mutex_unlock(&ctx->lock);
    return CAFS_OK;
}

File: Src/FUSE/Linux/x86-64/I-O_Engine/src/check.rs
//! `CafsCtx::check` — verifies the Superblock's BLAKE3-128 checksum.
//! On mismatch: tries Superblock Backup (via LBA0's fallback
//! pointer); if that's good, copies it over the primary in place
//! (when the mount is writable) and logs the repair. All other
//! anchor structures are neither checked nor repaired in V1
//! (ADR-007).
//!
//! SPDX-License-Identifier: BSD-3-Clause
//! Copyright (c) 2026, Imran Bin Gifary (System Delta)

use crate::block_io::{read_raw, write_raw};
use crate::checksum;
use crate::error::CafsResult;
use crate::layout::*;
use crate::{CafsCtx, CheckResult};

fn superblock_checksum_ok(sb: &[u8; REAL_BLOCK_4K as usize]) -> bool {
    let computed = checksum::blake3_128(&sb[0..SB_HASHED_LEN]);
    computed == sb[SB_OFF_CHECKSUM..SB_OFF_CHECKSUM + SB_CHECKSUM_LEN]
}

impl CafsCtx {
    pub fn check(&mut self) -> CafsResult<CheckResult> {
        let result = CheckResult {
            pointer_checksum_algo_used: Some(self.lba0_algo),
            ..Default::default()
        };

        if superblock_checksum_ok(&self.superblock) {
            result.superblock_ok = true;
            return Ok(result);
        }

        // Primary is bad — try the backup.
        let Some((backup_offset, _size)) = self.lba0.fallback else {
            eprintln!("cafs: primary Superblock checksum mismatch, and no backup pointer is set — unrecoverable");
            return Ok(result); // superblock_ok stays false
        };

        let mut backup = [0u8; REAL_BLOCK_4K as usize];
        if read_raw(self, backup_offset, &mut backup).is_err() || !superblock_checksum_ok(&backup) {
            eprintln!(
                "cafs: primary Superblock checksum mismatch, and Superblock Backup at offset {} is also bad or unreadable — unrecoverable",
                backup_offset.0
            );
            return Ok(result);
        }

        // Backup is good. Adopt it in memory regardless of write
        // access, so the rest of this session sees correct data even
        // on a readonly mount.
        self.superblock = backup;
        result.superblock_ok = true;

        if self.want_write {
            write_raw(self, self.lba0.real_offset, &self.superblock)?;
            eprintln!(
                "cafs: repaired primary Superblock at offset {} from Superblock Backup at offset {}",
                self.lba0.real_offset.0, backup_offset.0
            );
            result.superblock_repaired = true;
        } else {
            eprintln!(
                "cafs: primary Superblock checksum mismatch — valid backup found, but mount is readonly, so the fix was not written back"
            );
        }

        Ok(result)
    }
}
